Replace flatpak manifest with a BuildStream project

dylanmccall opened the pull request:
This project includes flatpak-runtimes.bst, which generates the
com.endlessm.apps.Platform runtime, along with Sdk runtime and
extensions.

Depends on https://github.com/endlessm/eos-knowledge-lib/pull/1700

https://phabricator.endlessm.com/T28284

dylanmccall committed:
Replace flatpak manifest with a BuildStream project

This project includes flatpak-runtimes.bst, which generates the
com.endlessm.apps.Platform runtime, along with Sdk runtime and
extensions. We track the sdk-6 git branch for all Endless SDK
components.

https://phabricator.endlessm.com/T28284

dylanmccall committed:
Add libpoppler with an additional bug fix

The version of libpoppler included with the GNOME 3.34 runtime has an
issue loading a PDF from an input stream, which is required in
eos-knowledge-lib.

https://phabricator.endlessm.com/T28284

dylanmccall commented:
I'm removing _.travis.yml_ in this PR for now since it definitely needs to be rewritten. The current one runs `make check`, and since we don't currently have a Makefile, that won't work :) All that check did was `json-glib-validate com.endlessm.apps.Sdk.json`. Could perhaps something similar with buildstream, running `bst show flatpak-runtimes.bst` or something along those lines? But need a docker container with bst available, first.

dylanmccall committed:
Remove central ref storage

Instead, we'll track refs directly in source declarations. This removes
a source of conflict with refs inside junctions, which will make it
easier to build the SDK.

https://phabricator.endlessm.com/T28284

dylanmccall committed:
Add elements for additional fonts from Google Fonts

Instead of maintaining our own zip files, font files are loaded from
the Google Fonts git repository. To save on disk space, the raw files
are downloaded directly from github.com instead of cloning the
repository.

https://phabricator.endlessm.com/T28284

dylanmccall commented:
For Google Fonts, there are a few ways to do this:

 * The very simple-looking way in this branch. Clone the https://github.com/google/fonts repo, read font files from there. Unfortunately, that repo is gigantic.
 * The uglier way that uses less disk space, as in this branch: https://github.com/endlessm/endless-sdk-flatpak/tree/dylanmccall/download-google-fonts-directly. I don't like it as much because it involves a giant file listing, but it is the most resource efficient.
 * The old way, maintaining zip files here. Would love to go with one of the first two since then we'll have a single source of truth for everything including fonts :)

wjt commented:
What is the issue with specifying URLs of the form `https://fonts.google.com/download?family=$FONT` in the `.bst` files?

dylanmccall commented:
The issue with doing that is buildstream generates a sha256sum of the contents, and those URLs aren't versioned - so if Google decides to update a font (or their zip file compression options), our build breaks. It isn't the _worst_ thing, but it could be a source of unhappy confusion in the future. But thinking about it, it counts as a legitimate inexpensive option - just needs a note in the README.

wjt commented:
Does the build break, or does it use the cached results of the last build?

dylanmccall commented:
To answer the question about breaking the build, I expect it would use cached results locally (buildstream won't try to fetch the same ref twice), but then break for someone doing a clean build. The worst kind of build failure :b

(Speaking of which, I just learned that's exactly why we had vendored the font files originally: https://github.com/endlessm/endless-sdk-flatpak/commit/c91e41c064c43cab4fb406e054b8a8a163e90ae6).

Following our chat earlier, I switched to using a list of versioned file URLs from GitHub, with a comment explaining where they came from. Listing all these files feels overly specific, but in some ways I find that reassuring. [Some future person](https://xkcd.com/1205/) might want to write a script or a buildstream github_files source plugin.

dylanmccall commented:
I was having trouble with a test in eos-knowledge-lib, [tests/js/framework/modules/window/testSimple.js](https://github.com/endlessm/eos-knowledge-lib/blob/dylanmccall-gnome-3-32/tests/js/framework/modules/window/testSimple.js), erroring out following a segfault inside GTK. The following seems to reproduce the crash:

```
const Endless = imports.gi.Endless;
const Gio = imports.gi.Gio;
const GLib = imports.gi.GLib;

let fake_pid = GLib.random_int();
let id_string = 'com.endlessm.knowledge.test.dummy' + GLib.get_real_time() + fake_pid;

let app = new Endless.Application({
    application_id: id_string,
    flags: 0
});

app.connect('startup', () => print("started"));
app.hold();

let window1 = new Endless.Window(app);
```

For SDK 6, the bottom of the stack trace looks like this:

```
#0  0x00007fffee3fb7a4 in _gtk_settings_get_screen (settings=0x0) at gtksettings.c:3319
#1  0x00007fffee2a3c3c in gtk_css_value_icon_theme_compute
    (icon_theme=<optimized out>, property_id=<optimized out>, provider=<optimized out>, style=<optimized out>, parent_style=<optimized out>) at gtkcssiconthemevalue.c:84
#2  0x00007fffee2c4027 in gtk_css_static_style_compute_value (style=
    0x55555598c050 [GtkCssStaticStyle], provider=0x0, parent_style=0x0, id=3, specified=0x7fffee825ba0 <inherit>, section=0x0) at gtkcssstaticstyle.c:237
#3  0x00007fffee2af242 in _gtk_css_lookup_resolve
    (lookup=lookup@entry=0x55555598a860, provider=provider@entry=0x0, style=style@entry=0x55555598c050 [GtkCssStaticStyle], parent_style=parent_style@entry=0x0) at gtkcsslookup.c:122
#4  0x00007fffee2c3f52 in gtk_css_static_style_new_compute
    (provider=provider@entry=0x0, matcher=matcher@entry=0x0, parent=parent@entry=0x0)
    at gtkcssstaticstyle.c:195
#5  0x00007fffee2c3fad in gtk_css_static_style_get_default () at gtkcssstaticstyle.c:164
#6  0x00007fffee2afc76 in gtk_css_node_init (cssnode=0x555555875ac0 [GtkCssNode]) at gtkcssnode.c:667
#7  0x00007ffff7cae161 in g_type_create_instance (type=<optimized out>) at ../gobject/gtype.c:1861
#8  0x00007ffff7c903a5 in g_object_new_internal
    (class=class@entry=0x555555878de0, params=params@entry=0x0, n_params=n_params@entry=0)
    at ../gobject/gobject.c:1827
#9  0x00007ffff7c91aa5 in g_object_new_with_properties
    (object_type=0x55555598a060 [GtkCssWidgetNode/GtkCssNode], n_properties=0, names=names@entry=0x0, values=values@entry=0x0) at ../gobject/gobject.c:1995
#10 0x00007ffff7c92651 in g_object_new
    (object_type=<optimized out>, first_property_name=first_property_name@entry=0x0)
    at ../gobject/gobject.c:1667
#11 0x00007fffee2cc97f in gtk_css_widget_node_new (widget=widget@entry=0x5555559884d0 [GtkWidget])
    at gtkcsswidgetnode.c:297
#12 0x00007fffee4aa07d in gtk_widget_init (instance=0x5555559884d0 [GtkWidget], g_class=0x55555587a440)
    at gtkwidget.c:4465
#13 0x00007ffff7cae161 in g_type_create_instance (type=<optimized out>) at ../gobject/gtype.c:1861
…
```

Similar crash in SDK 5. I'm weirded out by that, and suspect maybe our [CSS patch](https://github.com/endlessm/endless-sdk-flatpak/blob/master/gtk3-CSS-eos-cairo-filter-property.patch) but no idea so far really :/

dylanmccall committed:
Add EndlessOS icon theme extension

https://phabricator.endlessm.com/T28284

bochecha commented:
> But need a docker container with bst available, first.

You could use the freedesktop-sdk one?

https://gitlab.com/freedesktop-sdk/infrastructure/freedesktop-sdk-docker-images/container_registry

That's what we use in our CI over at freedesktop-sdk.

bochecha commented on elements/gnome-sdk-core-deps/poppler.bst:
```diff
@@ -0,0 +1,37 @@
+  url: https://poppler.freedesktop.org/poppler-0.72.0.tar.xz
+  ref: c1747eb8f26e9e753c4001ed951db2896edc1021b6d0f547a0bd2a27c30ada51
+- kind: patch
+  path: files/poppler/0001-glib-Don-t-create-PopplerInputStream-with-length-0.patch
```
This looks like an important patch. Did you try and talk to the GNOME release team? (who maintain the GNOME Sdk at [gnome-build-meta](https://gitlab.gnome.org/GNOME/gnome-build-meta/))

They might be interested in adding that patch to their element (they regularly add important patches until upstream makes a new release), at which point you can drop this entirely. :slightly_smiling_face:

dylanmccall replied:
Now you mention it, yeah! That patch is from here:

https://gitlab.freedesktop.org/poppler/poppler/merge_requests/189

It was released in poppler-0.78. Seems like they probably could get away with a version bump, but at the same time they don't actually include evince or poppler in the Flatpak Sdk upstream. It's probably a bit silly that I'm not just including our own evince and poppler elements, come to think of it. But I made a ticket over here: https://gitlab.gnome.org/GNOME/gnome-build-meta/issues/221. Indeed if their libevince gives us a working build that'll be *way* better :)

bochecha commented on elements/gnome-sdk-sdk/WebKitGTK.bst:
```diff
@@ -0,0 +1,87 @@
+# Replaces sdk/WebKitGTK.bst from gnome-sdk.
```
Ouch!

Do try hard to reuse the GNOME one, building this is going to waste so much time and computing power. :persevere: 

It seems the only difference is you're adding `-DENABLE_GLES2=ON` on ARM, GNOME might want to do the same?

dylanmccall replied:
Yeah, that's all we're doing here, which is a bit sad. At the moment it's just blindly following the original [generate-manifest.py](https://github.com/endlessm/endless-sdk-flatpak/blob/master/generate-manifest.py). I'll do some archaeology and then ask around on IRC :)

bochecha commented on elements/gnome-sdk-sdk/gst-plugins-good.bst:
```diff
@@ -0,0 +1,37 @@
+# Replaces sdk/gst-plugins-good.bst from gnome-sdk.
```
You're replacing this from both FDO and GNOME Sdks?

bochecha commented on elements/icon-theme-endless-os/appdata.bst:
```diff
@@ -0,0 +1,20 @@
+
+sources:
+- kind: local
+  path: files/icon-theme-endless-os-appdata
```
Any reason not to have that added to `eos-theme.bst`, and in fact just remove the `stack` element?

dylanmccall replied:
That would be _way_ better! Thanks :)

bochecha commented on elements/sdk-build-depends/cython.bst:
```diff
@@ -0,0 +1,9 @@
+sources:
+- kind: git_tag
+  url: github_com:cython/cython
+  track: 0.29.13
```
This seems like something we might want to add to the FDO Sdk. :thinking: 

Did you ask over there?

dylanmccall replied:
That makes sense! Feature request with a patch filed upstream: https://gitlab.com/freedesktop-sdk/freedesktop-sdk/issues/910

bochecha commented on elements/sdk-platform.bst:
```diff
@@ -0,0 +1,263 @@
+# There are certain elements inside gnome-sdk and freedesktop-sdk that we need
+# to build ourselves. As a result, we can't depend on gnome-sdk directly. This
+# list should be kept in sync with the dependencies in sdk-platform.bst from
+# gnome-sdk.
```
I think I'd add a link to the file in gnome-build-meta, for easier maintenance?

That would also make it clear which version of gnome-build-meta this is based upon.

I guess this is the file? https://gitlab.gnome.org/GNOME/gnome-build-meta/blob/gnome-3-34/elements/sdk-platform.bst

dylanmccall replied:
Good idea! Thank you :)

bochecha commented on elements/sdk/eos-sdk.bst:
```diff
@@ -0,0 +1,26 @@
+- kind: git_tag
+  url: github_com:endlessm/eos-sdk.git
+  track: sdk-6
+  ref: Release_3.3.18-56-g046c835870f14d53297435dc022740185c756cd1
```
This is a bit surprising…

On one hand you're saying you want to track the branch `sdk-6` but to only take tags from it (using the `git_tag` source).

On the other hand you explicitly take a commit which isn't on a tag (it's 56 commits after the `Release_3.3.18` tag).

Was that expected?

dylanmccall replied:
Yeah, git_tag by default tracks the latest commit - it only tracks the latest tag if you specify `track_tags: True`. (Which is kind of backwards, now that you mention it). This is the case for a lot of our dependencies, although it would be nice to track tags for repos that we know are good about tagging releases.

dylanmccall replied:
Ah, I see where that's weird. In gnome-build-meta, they stick with the default track-tags: False. In freedesktop-sdk, the default is track-tags: True. This one is closer to gnome-build-meta, albeit for no particular reason :)

bochecha reviewed (commented):
Overall this is looking pretty good, from a quick read. I only have a few questions, mostly to better understand some of the things you're doing here.

I haven't been able to test building it though, because bst complains:

```
[--:--:--][][] BUG     <built-in method keys of SanitizedDict object at 0x7f4df2e71f80> returned a result with an error set

    TypeError: �    ��M is not JSON serializable
    
    The above exception was the direct cause of the following exception:
    
    SystemError: <built-in method keys of SanitizedDict object at 0x7f4df7c88b00> returned a result with an error set
    
    The above exception was the direct cause of the following exception:
    
    SystemError: <built-in method keys of SanitizedDict object at 0x7f4df2e6edd0> returned a result with an error set
    
    The above exception was the direct cause of the following exception:
    
    Traceback (most recent call last):
      File "/usr/bin/bst", line 8, in <module>
        sys.exit(cli())
      File "/home/mathieu/.local/lib/python3.7/site-packages/click/core.py", line 764, in __call__
        return self.main(*args, **kwargs)
      File "/usr/lib/python3.7/site-packages/buildstream/_frontend/cli.py", line 174, in override_main
        standalone_mode=standalone_mode, **extra)
      File "/home/mathieu/.local/lib/python3.7/site-packages/click/core.py", line 717, in main
        rv = self.invoke(ctx)
      File "/home/mathieu/.local/lib/python3.7/site-packages/click/core.py", line 1137, in invoke
        return _process_result(sub_ctx.command.invoke(sub_ctx))
      File "/home/mathieu/.local/lib/python3.7/site-packages/click/core.py", line 956, in invoke
        return ctx.invoke(self.callback, **ctx.params)
      File "/home/mathieu/.local/lib/python3.7/site-packages/click/core.py", line 555, in invoke
        return callback(*args, **kwargs)
      File "/home/mathieu/.local/lib/python3.7/site-packages/click/decorators.py", line 27, in new_func
        return f(get_current_context().obj, *args, **kwargs)
      File "/usr/lib/python3.7/site-packages/buildstream/_frontend/cli.py", line 344, in build
        build_all=all_)
      File "/usr/lib/python3.7/site-packages/buildstream/_stream.py", line 182, in build
        dynamic_plan=True)
      File "/usr/lib/python3.7/site-packages/buildstream/_stream.py", line 945, in _load
        self._pipeline.resolve_elements(elements)
      File "/usr/lib/python3.7/site-packages/buildstream/_pipeline.py", line 139, in resolve_elements
        element._update_state()
      File "/usr/lib/python3.7/site-packages/buildstream/element.py", line 1091, in _update_state
        self.__weak_cache_key = self.__calculate_cache_key(dependencies)
      File "/usr/lib/python3.7/site-packages/buildstream/element.py", line 2007, in __calculate_cache_key
        return _cachekey.generate_key(cache_key_dict)
      File "/usr/lib/python3.7/site-packages/buildstream/_cachekey.py", line 42, in generate_key
        ustring = ujson.dumps(ordered, sort_keys=True, escape_forward_slashes=False).encode('utf-8')
    SystemError: <built-in method keys of SanitizedDict object at 0x7f4df2e71f80> returned a result with an error set
```

I need to investigate that…

dylanmccall committed:
Track eos-knowledge-lib branch dylanmccall-gnome-3-32

This makes it easier to test the runtime, but should be reverted before
releasing.

https://phabricator.endlessm.com/T28284

dylanmccall committed:
Add a comment pointing to upstream sdk-platform.bst

https://phabricator.endlessm.com/T28284

dylanmccall commented:
Thank you for looking through this so thoroughly @bochecha! And that Docker container looks perfect. I'll try it :)

@dnohales and I both had that same issue with buildstream (although from system-installed site-packages/click, rather than user-installed). At this point I'm using bst 1.4.1 from Fedora's repositories and it works happily, but earlier on I'd worked around that ugly crash by using a version I had installed with pip. It'll be good to see what happens when I run it with the Docker container.

dylanmccall committed:
Use poppler from the newest version of gnome-sdk

https://phabricator.endlessm.com/T28284

dylanmccall committed:
Simplify structure around the icon theme extension

https://phabricator.endlessm.com/T28284

bochecha commented:
> @dnohales and I both had that same issue with buildstream (although from system-installed site-packages/click, rather than user-installed). At this point I'm using bst 1.4.1 from Fedora's repositories and it works happily

Good to know I don't maintain that package for nothing. :sweat_smile: 

The error is in fact an incompatibility between the ujson 2.0 in the Fedora repositories and bst-external: https://gitlab.com/BuildStream/bst-external/merge_requests/109

If you install bst-external [from my copr](https://copr.fedorainfracloud.org/coprs/bochecha/buildstream/), it includes the patch and won't have this problem. Otherwise if you install bst with Pip, you also don't get the problem because then ujson 1.35 is installed with which bst-external is compatible.

I managed to build this Sdk after fixing this issue, I'll try to have a look at it in the future to see if some things can be improved, but that can come in a future merge request. :slightly_smiling_face: 

Out of curiosity, did you have a look at the ABI checker I wrote for the Freedesktop Sdk? You might want something like that to catch some ABI breaks and avoid breaking your apps. :slightly_smiling_face:

bochecha commented on elements/extensions/icon-theme-endless-os.bst:
```diff
@@ -0,0 +1,35 @@
+- gnome-sdk.bst:sdk/appstream-glib.bst
+- gnome-sdk.bst:sdk/glib.bst
+- gnome-sdk.bst:sdk/librsvg.bst
+
```
I think you should at least add a runtime dependency on Gtk3, otherwise there is a risk this integration command gets run **before** staging the icon theme when composing the final Flatpak image:

https://gitlab.gnome.org/GNOME/gnome-build-meta/blob/gnome-3-34/elements/sdk/gtk+-3.bst#L48-54

You can either add the following:

```
runtime-depends:
- gnome-sdk-sdk/gtk+-3.bst
```

Or you can do the following change:

```diff
 build-depends:
 - freedesktop-sdk.bst:public-stacks/buildsystem-autotools.bst
-- gnome-sdk-sdk/gtk+-3.bst
 - gnome-sdk.bst:core-deps/intltool.bst
 - gnome-sdk.bst:sdk/appstream-glib.bst
 - gnome-sdk.bst:sdk/glib.bst
 - gnome-sdk.bst:sdk/librsvg.bst

+depends:
+- gnome-sdk-sdk/gtk+-3.bst
+
```

dylanmccall replied:
So if I do that the extension includes a bunch of gtk+-3 related stuff (such as /usr/share/runtime/share/icons/EndlessOS/etc/…), although we can strip that out if need be. I don't quite understand what this solves. Would it help if we do our own gtk-update-icon-cache here?

bochecha replied:
> I don't quite understand what this solves.

You need to have the icon theme present in the root before running `gtk-update-icon-cache`, and I **think** you can only guarantee that if the icon theme `depends` on Gtk.

However, it seems I was wrong because in FDO Sdk Hicolor doesn't depend on Gtk: https://gitlab.com/freedesktop-sdk/freedesktop-sdk/blob/master/elements/components/hicolor-icon-theme.bst

dylanmccall committed:
Use upstream WebKitGTK.bst

dylanmccall committed:
Pass arch variable to gnome-sdk junction

dylanmccall committed:
Add -Wno-error=cast-align for xapian-core

dylanmccall committed:
Add .travis.yml, and a new Makefile

https://phabricator.endlessm.com/T28284

dylanmccall committed:
Update README.md with some additional information

The README now includes some help for installing BuildStream, if needed,
and information about building for 32-bit ARM.

https://phabricator.endlessm.com/T28284

dylanmccall committed:
Fetch dependencies before bst show

https://phabricator.endlessm.com/T28284

dylanmccall committed:
Output a single repo in Makefile, as before

In addition, `make bundle` produces a collection of Flatpak bundle
files, which can be helpful for testing.

https://phabricator.endlessm.com/T28284

dylanmccall committed:
Version bump for eos-knowledge-lib

dylanmccall commented:
Oh, update on Travis CI: In the past few commits I put back .travis.yml and a rudimentary makefile with similar behaviour to the old one (minus some targets we don't need any more, and with a handy new _bundle_ target which eats up piles of disk space and spits out a directory full of flatpak bundle files). Travis just runs `make check`, which itself runs `bst show` on our project files to confirm they are valid.

One detail to note with the makefile: our buildstream project generates a brand new repo for both flatpak-runtimes.bst and flatpak-platform-extensions.bst, and the makefile pulls those together itself into a third repo by running `ostree pull-local`. This way the makefile behaves the same as the old one, and we can still output to a pre-existing ostree repo. It does take up more space, though. Those superfluous repos go into a directory called _./cache_ (or whatever _CACHEDIR_ is set to), and can be safely deleted.

dylanmccall committed:
Add version information to runtime builds

This is passed to sdk/os-release.bst by way of a yml file which is
included by the element. Unfortunately, this implementation requires
that we add a file (flatpak-version.yml) which is tracked by git.

This might be tidier if buildstream allows us to add local variable
overrides, for example in a local project.conf.

https://phabricator.endlessm.com/T28284

ptomato committed:
Add EndlessOS instructions in README

Adds a few instructions on what dependencies to install and how to patch
BuildStream so it works with FUSE 3.x.

dylanmccall committed:
Improve how we track git_tag sources copied from junctions

Instead of keeping the existing track and ref properties for git_tag
sources that are copied from gnome-sdk.bst and freedesktop-sdk.bst,
change "ref" properties to "track" with the equivalent commit hash. This
ensures that our version will always point to the same commit as the
upstream version, even after re-running `bst track`.

https://phabricator.endlessm.com/T28284

dylanmccall committed:
Update refs for sdk sources

https://phabricator.endlessm.com/T28284

dylanmccall committed:
Install tcl man pages with a different suffix

This fixes a problem with files in sdk-build-depends/tcl.bst overlapping
freedesktop-sdk.bst:components/perl.bst.

https://phabricator.endlessm.com/T28284

dylanmccall committed:
Prevent accidental changes to flatpak-version.yml

https://phabricator.endlessm.com/T28284

dylanmccall committed:
Force non-interactive mode for bst in Makefile

https://phabricator.endlessm.com/T28284

dylanmccall committed:
Update refs and track same freedesktop-sdk as gnome-sdk

Previously we were tracking a slightly different version of
freedesktop-sdk than the version used by gnome-sdk, which could result
in subtle problems.

https://phabricator.endlessm.com/T28284

dylanmccall committed:
Add aarch64 build config for gtk+-3.bst

https://phabricator.endlessm.com/T27755

dylanmccall committed:
Add a make target to sign ostree commits

https://phabricator.endlessm.com/T28284

dylanmccall committed:
Update release process in README.md

https://phabricator.endlessm.com/T28284

dylanmccall committed:
Use build.conf with Makefile builds

https://phabricator.endlessm.com/T28284

dylanmccall committed:
Track newest versions of eos-sdk and eos-knowledge-lib

https://phabricator.endlessm.com/T28284

dylanmccall commented:
Unless there are objections, I'm planning to go ahead and merge this into the sdk-6 branch so we can get Jenkins building it in endlessm/eos-build#1068. Will create another pull request later to merge sdk-6 back into master.

dbnicholson commented on Makefile:
```diff
@@ -1,149 +1,152 @@
-	$(call build-manifest,com.endlessm.apps.Sdk.json)
+define sign-ref
+	$(shell ostree gpg-sign --repo=$(1) $(2) $(GPG_KEY_ID))
+endef
```
Rather than gathering the flatpak refs yourself and calling `ostree gpg-sign`, you can just call `flatpak build-sign --gpg-sign $(GPG_KEY_ID) $(REPO)`. If you don't supply a flatpak ID, it will loop through the repo looking for all `app/` and `runtime/` refs to sign.

dylanmccall replied:
:o that is so much better. Thank you! I changed this to use `flatpak build-sign` and `flatpak build-update-repo` like you demonstrated and removed a bunch of gunk in the process. Indeed, the _sign_ target is unnecessary with that change.

dbnicholson commented on Makefile:
```diff
@@ -1,149 +1,152 @@
+	mkdir -p $(CACHEDIR)
+
+$(REPO):
+	$(OSTREE) init --repo=$(REPO) --mode=bare-user-only
```
You need this to remain an archive repo or we won't be able to publish from it. I suppose you could precreate the repo in the jenkins script the way we need it there. OSTree just returns immediately if the repo is already initialized.

dbnicholson commented on Makefile:
```diff
@@ -1,149 +1,152 @@
+
+EXPORT-$(FLATPAK_RUNTIMES_REPO): $(FLATPAK_RUNTIMES_REPO) | $(REPO)
+	$(OSTREE) pull-local --repo=$(REPO) $(FLATPAK_RUNTIMES_REPO)
+	$(OSTREE) summary --repo=$(REPO) --update
```
You'll also need to sign the ostree summary in order to publish it. Maybe you want to just keep this and the commit signing in the generic `export` target, keep the `EXPORT_ARGS` as before and make the commit signing conditional on it being set. Something like this:
```
EXPORT_ARGS ?=
export:
ifneq ($(EXPORT_ARGS),)
	flatpak build-sign $(EXPORT_ARGS) $(REPO)
endif
	flatpak build-update-repo $(EXPORT_ARGS) $(REPO)
```
`EXPORT_ARGS` really means `GPG_SIGN_ARGS`, but this is just using the same name that the jenkins script uses. One advantage of just using a bare variable is that it allows passing in a `--gpg-homedir` if necessary. Currently we don't do that, but with your changes it wouldn't be possible.

Then all the `EXPORT-*` targets would just do `pull-local` as prereqs for the generic `export` target.

Or, another way to manage the keys and homedir:
```
GPG_KEY_IDS ?=
GPG_HOMEDIR ?=
GPG_ARGS = $(strip $(patsubst %,--gpg-sign=%,$(GPG_KEY_IDS)) \
	$(patsubst %,--gpg-homedir=%,$(GPG_HOMEDIR)))
export:
ifneq ($(GPG_ARGS),)
	flatpak build-sign $(GPG_ARGS) $(REPO)
endif
	flatpak build-update-repo $(GPG_ARGS) $(REPO)
```

dylanmccall replied:
Okay, fixed that now. I went with the first approach :) Thank you for the detailed explanation!

dylanmccall committed:
Use flatpak build commands to sign export repo

https://phabricator.endlessm.com/T28284

dylanmccall committed:
Use --mode=archive for the export repo

https://phabricator.endlessm.com/T28284

dbnicholson commented:
For reference, the jenkins side of this is https://github.com/endlessm/eos-build/pull/1068.

Shouldn't this PR be targeting master and then we'd merge/fast-forward master back to sdk-6? I don't think we want to keep master building against the old GNOME SDK. At this point (IMO), any work on the stable SDK should be going on the sdk-5 branch and development work goes on master.

dylanmccall commented:
Yeah, that's true. The master branch was suddenly looking very active, but that was just the aarch change. Thinking about it again, makes much more sense to just say future changes to sdk-5 happen in that branch :)

dbnicholson commented on elements/sdk/eos-knowledge-lib.bst:
```diff
@@ -0,0 +1,38 @@
+sources:
+- kind: git_tag
+  url: github_com:endlessm/eos-knowledge-lib.git
+  track: sdk-6
```
This git branch seems to be used in lots of places. Perhaps it could be a variable? That would make bumping to the next release much easier.

dylanmccall replied:
Yeah, good call. We can add that to project.conf - just can't be a build-time option.

dylanmccall replied:
Now I remember, I tried doing that but the git_tag plugin doesn't resolve variable references there. (I imagine BuildStream is employing a large hammer here to avoid some mishaps, but maybe they'd be cool making the track field do variable substitution...).

dbnicholson commented on flatpak-version.yml:
```diff
@@ -0,0 +1,7 @@
+# no-commit
+# This file is generated by utils/generate-version as part of the
+# Makefile / CI system. Please do not commit it.
```
Looks like you did commit it.

dylanmccall replied:
:man_facepalming: curse you not running my pre-commit hooks, gitg!

dbnicholson commented on project.conf:
```diff
@@ -0,0 +1,306 @@
+  - '${XDG_RUNTIME_DIR}'
+
+variables:
+  sdk-branch: "6"
```
Should this be `master` on master? Currently when you build master it creates `runtime/com.endlessm.apps.Sdk/*/master`.

dylanmccall replied:
Okay, changed this to master, and changed a bunch of other elements to track the master branches of their respective repos. When this is merged we can add a commit in the sdk-6 branch that tracks sdk-6 branches instead.

dbnicholson commented on elements/sdk/eos-knowledge-lib.bst:
```diff
@@ -0,0 +1,38 @@
+- kind: git_tag
+  url: github_com:endlessm/eos-knowledge-lib.git
+  track: sdk-6
+  ref: Release_3.2.6-699-g7de89777bf721bd257e2b9240401d18f8a8add12
```
Since we don't actually tag our SDK dependencies, does it make sense to specify this? Clearly this is just random `git-describe` output and not an actual tag. https://gitlab.gnome.org/GNOME/gnome-build-meta/blob/master/elements/sdk/glib.bst for instance doesn't specify `ref`, presumably because they want to just use whatever is the latest on `master`.

I actually don't really understand why the `git_tag` element is preferred over the regular `git` element. Both seem to handle git tags. @bochecha?

dylanmccall replied:
The idea is a particular commit in endless-sdk-flatpak is tracking a particular commit in each repo, and that ref changes (according to the value of the _track_ property) whenever we run `bst track`. It is possible to track at build time (using a separate project.refs file which is gitignored). That's what gnome-build-meta uses for their _master_ branch. For release branches they use either the same format as us, or use download URLs instead of git commits.

Good to ask about git vs git_tag since we're definitely using it here for more than tracking tags, which is peculiar :) I can't fully recall the reason here. For a moment I thought this was due to git_tag having a _track_ property, but the usual git plugin has that as well. To some extent it feels nice to me using the same source plugin consistent, though, even if we're only using its features in a couple places.

dylanmccall replied:
On tracking at build time, I predict that we may want to discuss that at some point, but for the small number of extra pieces in our SDK I imagine we'll be fine without as long as we have some replacement for the per-component "flapjack-" jobs.

dylanmccall committed:
Update refs and track tags for eos-theme

https://phabricator.endlessm.com/T28284

dylanmccall committed:
Track the master branch for SDK component repos

Tracking sdk-6 should only happen with builds created from the sdk-6
branch of this repository.

https://phabricator.endlessm.com/T28284
